apis/v1alpha2/tcproute_types.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=gateway-api
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// TCPRoute is the Schema for the TCPRoute resource.
+type TCPRoute struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of TCPRoute.
+	Spec TCPRouteSpec `json:"spec"`
+
+	// Status defines the current state of TCPRoute.
+	Status TCPRouteStatus `json:"status,omitempty"`
+}
+
+// TCPRouteSpec defines the desired state of TCPRoute
+type TCPRouteSpec struct {
+	CommonRouteSpec `json:",inline"`
+
+	// Rules are a list of TCP matchers and actions.
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	Rules []TCPRouteRule `json:"rules"`
+}
+
+// TCPRouteStatus defines the observed state of TCPRoute
+type TCPRouteStatus struct {
+	RouteStatus `json:",inline"`
+}
+
+// TCPRouteRule is the configuration for a given rule.
+type TCPRouteRule struct {
+	// Matches define conditions used for matching the rule against incoming TCP
+	// connections. Each match is independent, i.e. this rule will be matched if
+	// **any** one of the matches is satisfied. If unspecified (i.e. empty),
+	// this Rule will match all requests for the associated Listener.
+	//
+	// Each client request MUST map to a maximum of one route rule. If a request
+	// matches multiple rules, matching precedence MUST be determined in order
+	// of the following criteria, continuing on ties:
+	//
+	// * The most specific match specified by ExtensionRef. Each implementation
+	//   that supports ExtensionRef may have different ways of determining the
+	//   specificity of the referenced extension.
+	//
+	// If ties still exist across multiple Routes, matching precedence MUST be
+	// determined in order of the following criteria, continuing on ties:
+	//
+	// * The oldest Route based on creation timestamp. For example, a Route with
```

Comment:
So a user who want to change the precdeance must delete the current `older` then create a new rule then add the `older` deleted rule? if so, then this is not a good a UX. Let us not infer precedence. User can add weight (if spec then we need to resolve a situation when two rules have the same weight) or even a `map[weight] match` where duplicate weight can not happen.

Reply:
Yeah this is really not a situation we want to ever end up in. The only reason you should end up in this situation is because of some kind of unfortunate mistake. In that case, we were left with 2 options:

1. Try to determine the Route that was attached to the Gateway first. Although this would be ideal, it requires statefulness and this knowledge could easily be lost if a controller restarted.
2. Consider this an error state and not support either Route until it was resolved.
3. Assume the older Route was attached first. This is not always going to be true, but the behavior is consistent, and it generally fulfills the goal of "prefer not to break things that are working".

If we were to let a user add weight here, how would we do that? Would it have to be centralized on the Gateway?

Reply:
Yeah this is a good point, we need to document this better. I think the most common scenario for multi namespace Gateways is still going to involve single-namespace listeners. So a foo.com listener trusts routes from the foo namespace and a bar.com listener trusts routes from the bar namespace.